### ISSUE TITLE
Setpoint sends every change to the homematic binding

### ISSUE THREAD

**Issue by patrickse** [openhab/openhab#1903]
Hi,

I think the subject for this issue is not the best. What I try to say is, that I´ve implemented a setpoint item for the target temperature in my living room. Everytime I hit the +/- it directly forwards this command to my valves. I defined 0.5° steps and my problem is, that when I switch from 17° to 20° it will update the valves from 17°->17.5°->18°->18.5° and so on. So I send a lot of traffic to my valves.

Is there a better solution or maybe a timeout for the setpoint control, to tell the control to just update if there´s no further change in about 10 secs or so.

Hope you got it... ;) 

Patrick

**Comment by hoalex** [comment]
I had the exact same thought when I was implementing my heating control in openHAB a year ago (using the FHT binding)... every change would be transmitted, which would cause a lot of traffic on the CUL stick.

So, what I ended up doing was to implement an additional "proxy item" that acts as a kind of gateway between the user input in the UI and the actual openHAB item where the rules are performed on.

My sitemap looks like this:
```
Setpoint item=px_Livingroom label="Desired temperature [%.0f °C]" icon="heating" step=1
Setpoint item=px_Kitchen label="Desired temperature [%.0f °C]" icon="heating" step=1
```

and my *.items file as follows:
```
Number  FHT_Livingroom_Desired  "Desired temperature Livingroom [%d °C]" <temperature>   (GF_Livingroom)             { fht="housecode=xxxx;datapoint=DESIRED_TEMP" }
Number  FHT_Kitchen_Desired  "Desired temperature Kitchen [%d °C]" <temperature>   (GF_Kitchen)             { fht="housecode=yyyy;datapoint=DESIRED_TEMP" }
Number px_Livingroom "Proxy Livingroom"
Number px_Kitchen "Proxy Kitchen"
```

As you can see, the FHT binding is defined in the FHT_* items, but the setpoints in the sitemap map to the proxy items.

Now, how to only send the last value the user set in set setpoint to the FHT? We implement a timer that is resetted every time a new update is received at the proxy variable!

My *.rules file:
```
var Timer timer = null

rule "change heating Livingroom"
when
        Item px_Livingroom received command
then
        if (timer != null)
                timer.cancel()

        timer = createTimer(now.plusSeconds(10)) [|
                        {
                                sendCommand(FHT_Livingroom_Desired, px_Livingroom.state.toString)
                        }
                ]
end
```
This way, every change of the setpoint that happens less than 10 seconds after the previous change, resets the timer to 10 seconds. If the user does not change the setpoint value any further, the timer runs out and sends the last received value to the actual item containing the FHT binding.

I'm using this for some time now and it works like a charm :-)

I hope this helped you.

Regards,
Alex

**Comment by patrickse** [comment]
Sounds really interesting. I think this could fix my problems.. I will give it a try this weekend...

Thanks for that .. :)

**Comment by cyclingengineer** [comment]
I had this issue when I was developing the Max! CUL binding. I noticed that
the wall controllers in the Max! system implemented something like your
rule does - waits for some seconds of no activity before sending the data.
Instead of making everyone implement rules as you have done (which will
work) I implemented it directly in the binding. I think this would be the
most desirable way to resolve your issue as otherwise everyone will have
the same problem and it makes your rules less clean :-)

See
https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.maxcul/src/main/java/org/openhab/binding/maxcul/internal/MaxCulBinding.java
and search for PACED_TRANSMIT_TIME

Hope that helps :)

**Comment by hoalex** [comment]
I agree with you that it's probably the best to implement this functionality directly in the individual bindings. So nice job there! :-)

However, this is not implemented in the FHT or HomeMatic bindings (as far as I know). So another way to handle this could be an additional parameter for the setpoint definition in a sitemap, e.g. "Setpoint item=Temperature_1 [...] delay=1000" for a delay of 1000 ms. That way, the scheduling would be handled directly by the openHAB core, and not every binding developer would have to implement this on their own. What do you think?

**Comment by patrickse** [comment]
It should be an optional parameter inside the sitemap and there should also be a default delay value set, inside the binding. So without further knowledge of the binding internals, the setpoint control should not send every click to the homematic devices. This will save battery and radio traffic.

Otherwise, if you´re using a non homematic device, it could be important to have a just in time update of the stepping value. So the binding should decide what is appropriate and give the user the choice to override this proposal by the sitemap´s delay parameter of the setpoint control.

**Comment by teichsta** [comment]
@gerrieg what do you think?

**Comment by gerrieg** [comment]
OK, i will implement a configurable delay for that special datapoint in the Homematic binding. Is there another Homematic datapoint which should be delayed too?

**Comment by patrickse** [comment]
Great news... looking forward and thanks for your efforts ...

**Comment by patrickse** [comment]
Perfect :+1:  Thank you

**Comment by teichsta** [comment]
issue has been implemented with #1952 - Thanks a lot @gerrieg for this contribution!
